core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -255,17 +294,182 @@ private void rebuildIndex() throws Exception {
       }
    }
 
+   // Initialise missing internal metadata state for corrupt data
+   private PrivateMetadata generateMissingInternalMetadata(K key) {
+      // Optimistic Transactions
+      AdvancedCache<?, ?> cache = ctx.getCache().getAdvancedCache();
+      ComponentRegistry componentRegistry = cache.getComponentRegistry();
+      Configuration config = cache.getCacheConfiguration();
+      TransactionConfiguration txConfig = config.transaction();
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
+      if (txConfig.transactionMode() == TransactionMode.TRANSACTIONAL && txConfig.lockingMode() == LockingMode.OPTIMISTIC) {
+         builder.entryVersion(new NumericVersionGenerator().generateNew());
+      }
+
+      // Async XSite
+      if (config.sites().hasAsyncEnabledBackups()) {
+         // TODO avoid cyclic dependency
+         KeyPartitioner keyPartitioner = componentRegistry.getComponent(KeyPartitioner.class);
+         IracVersionGenerator ivg = componentRegistry.getIracVersionGenerator().running();
+         int segment = keyPartitioner.getSegment(key);
+         builder.iracMetadata(ivg.generateMetadataWithCurrentVersion(segment));
+      }
+      return builder.build();
+   }
+
+   private void migrateFromV12_0()  throws Exception {
+      // ISPN-13128 Corrupt migration data can only be created with default marshaller
+      if (ctx.getGlobalConfiguration().serialization().marshaller() == null) {
+         migrateCorruptDataV12_0();
+         return;
+      }
+
+      // Data is not corrupt, so simply update magic and transfer file contents
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startMigratingPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+      if (newFile.exists()) {
+         newFile.delete();
+      }
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
```

Comment:
I guess if the "magic value" has the same length, you can update the old file directly instead of copying everything to a new file.